### PR TITLE
closes GROK-12842: Bio | ST: port hardcoded data to external sources

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: packages/SequenceTranslator/detectors.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  SOURCE: 'Source',
+  ICD: 'ICD',
+  OWNER: 'Owner',
+};
+
+class SequenceTranslatorPackageDetectors extends DG.Package {
+  //tags: semTypeDetector
+  //input: column col
+  //output: string semType
+  detectNucleotide(col) {
```
Comment (skalkin) [inline packages/SequenceTranslator/detectors.js:35]: Why is this needed?

File: packages/SequenceTranslator/detectors.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  //name: autostart
+  //tags: autostart
+  //description: SequenceTranslator bootstrap
+  autostart() {
+    console.debug('ST: detectors.js: autostart()');
+
+    const eventsOnViewAdded = (view) => {
+      console.debug('ST: detectors.js: eventsOnViewAdded() handler start');
```
Comment (skalkin) [inline packages/SequenceTranslator/detectors.js:52]: Please remove or comment out debug printouts

File: packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+const AXOLABS_STYLE_FILENAME = 'axolabs-style.json';
+
+export class JsonLoader {
+  private constructor() { }
+
+  private static instance?: JsonLoader;
+  private formatDictionary: FormatDict;
+  private axolabsStyle: AxolabsStyle;
+
+  static getInstance() {
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts:37]: Why the instance can't simply be initialized and used, what's getInstance for?

Also, does it have to be a class, can't it be a simple function?

File: packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    } catch (err: any) {
+      const errMsg: string = err.hasOwnProperty('message') ? err.message : err.toString();
+      throw new Error('ST: Loading format dictionary error: ' + errMsg);
+    }
+  }
+
+  public getFormatDictionary(): FormatDict {
+    if (this.formatDictionary === undefined)
+      throw new Error('ST: format dictionary not initialized');
+    return this.formatDictionary as FormatDict;
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts:61]: What is this cast for?

File: packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  private axolabsStyle: AxolabsStyle;
+
+  static getInstance() {
+    if (JsonLoader.instance === undefined)
+      JsonLoader.instance = new JsonLoader();
+    return JsonLoader.instance;
+  }
+
+  async init() {
+    try {
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts:44]: Why have a static sync getInstance method if it appears that it requires some form of async initialization?

File: packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+      const fileSource = new DG.FileSource(APP_PATH);
+
+      const formatDictFile = await fileSource.readAsText(FORMAT_DICT_FILENAME);
+      this.formatDictionary = JSON.parse(formatDictFile);
+
+      const axolabsStyleFile = await fileSource.readAsText(AXOLABS_STYLE_FILENAME);
+      this.axolabsStyle = JSON.parse(axolabsStyleFile);
+    } catch (err: any) {
+      const errMsg: string = err.hasOwnProperty('message') ? err.message : err.toString();
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts:53]: I've seen this code before, let's reuse the handling

File: packages/SequenceTranslator/src/model/format-translation/conversion-utils.ts [hunk 7]
```diff
@@ -0,0 +1,66 @@
+import {SYNTHESIZERS as FORMAT, TECHNOLOGIES} from '../const';
+import {FormatConverter} from './format-converter';
+
+const NO_TRANSLATION_MSG = 'No translation table available';
+export const UNDEFINED_SEQ_MSG = 'Type of input sequence is undefined';
+
+export function convertSequence(sequence: string, indexOfFirstInvalidChar: number, format: string | null): {[key: string]: string} {
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/format-translation/conversion-utils.ts:7]: Shouldn't format be enum?

File: packages/SequenceTranslator/src/model/format-translation/conversion-utils.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  if (indexOfFirstInvalidChar !== -1) {
+    return {
+      indexOfFirstInvalidChar: indexOfFirstInvalidChar.toString(),
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/format-translation/conversion-utils.ts:10]: Why would you store a number as a string?

File: packages/SequenceTranslator/src/model/format-translation/conversion-utils.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      Error: UNDEFINED_SEQ_MSG,
+    };
+  }
+  if (format === FORMAT.NUCLEOTIDES ) {
+    const converter = new FormatConverter(sequence, FORMAT.NUCLEOTIDES);
+    return {
+      type: FORMAT.NUCLEOTIDES,
+      Nucleotides: sequence,
+      BioSpring: converter.convert(FORMAT.BIOSPRING),
+      GCRS: converter.convert(FORMAT.GCRS),
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/format-translation/conversion-utils.ts:20]: GCRS format should not be public

File: packages/SequenceTranslator/src/model/registration/add-columns.ts [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    });
+  }
+
+  private addBatchMolWeightCol(): void {
+    const compoundMolWeightCol = this.df.getCol(COL_NAMES.COMPOUND_MOL_WEIGHT);
+    const saltMassCol = this.df.getCol(COL_NAMES.SALT_MASS);
+    this.df.columns.addNewFloat(COL_NAMES.BATCH_MOL_WEIGHT).init((i: number) => {
+      let res: number = Number.NaN;
+      try {
+        res = getBatchMolWeight(compoundMolWeightCol, saltMassCol, i);
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/registration/add-columns.ts:149]: Looks like duplicated code (see previous 20 lines). Too verbose.

File: packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts (hunk 4 above)
Comment (skalkin) [reply packages/SequenceTranslator/src/model/data-loading-utils/json-loader.ts:61]: Usually there is a reason why I ask questions, but I see that you often just straight answer the question :) this.formatDictionary is already declared as FormatDict, so why cast it?

File: packages/Bio/src/utils/monomer-lib.ts [hunk 11]
```diff
@@ -135,30 +143,27 @@ export class MonomerLibHelper implements IMonomerLibHelper {
...
-        'symbol': monomer['symbol'],
-        'name': monomer['name'],
-        'naturalAnalog': monomer['naturalAnalog'],
-        'molfile': monomer['molfile'],
-        'rgroups': monomer['rgroups'],
-        'polymerType': monomer['polymerType'],
-        'monomerType': monomer['monomerType'],
-        'data': {}
-      };
+      const monomerAdd: {[key: string]: any} = {};
```
Comment (LaykDimon) [inline packages/Bio/src/utils/monomer-lib.ts:147]: can't we make it typed?
any isn't a good approach

File: packages/Bio/src/utils/monomer-lib.ts [hunk 12]
```diff
@@ -135,30 +143,27 @@ export class MonomerLibHelper implements IMonomerLibHelper {
...
+      for (const field of HELM_REQUIRED_FIELDS_ARRAY)
+        monomerAdd[field] = monomer[field];
+      for (const optField of HELM_OPTIONAL_FIELDS_ARRAY) {
+        const value = monomer[optField];
+        if (value !== undefined)
```
Comment (LaykDimon) [inline packages/Bio/src/utils/monomer-lib.ts:152]: you can change it to `if (value)`

File: packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+export function findDuplicates(data: Int32Array | Float32Array | Float64Array | Uint32Array): number[] {
+  return Array.from(new Set(data)).filter((value) => data.indexOf(value) !== data.lastIndexOf(value));
+}
+
+export async function isCurrentUserCreatedThisPattern(patternName: string): Promise<boolean> {
+  return await grok.dapi.users.current().then((user) => {
+    const [firstName, lastName] = getUserName(patternName);
+    return (user.firstName != firstName || user.lastName != lastName);
```
Comment (LaykDimon) [inline packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts:24]: use strict comparison

File: packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+  });
+}
+
+export function getShortName(patternName: string): string {
+  let first = patternName.length + 1;
+  for (let i = 0; i < patternName.length; i++) {
+    if (patternName[i] == '(') {
```
Comment (LaykDimon) [inline packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts:31]: same here

File: packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+function getUserName(patternName: string): string[] {
+  let first = -1;
+  for (let i = 0; i < patternName.length; i++) {
+    if (patternName[i] == '(') {
+      first = i;
+      break;
+    }
+  }
+  return (first == -1) ? ['', ''] : patternName.slice(first + 9, patternName.length - 1).split(' ').slice(1);
```
Comment (LaykDimon) [inline packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts:47]: you need to put this 9 into const

File: packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    i++;
+    const AXOLABS_MAP = axolabsStyleMap;
+
+    const baseChoices: string[] = Object.keys(AXOLABS_MAP);
+    // const defaultBase: string = baseChoices[0];
+    const indexOfSymbol = AXOLABS_MAP['RNA']['symbols'].indexOf(x);
+    let symbol = AXOLABS_MAP[bases[i].value]['symbols'][indexOfSymbol];
+    if (isOverhang(bases[i].value)) {
+      if (i < sequence.length / 2 && !isOverhang(bases[i + 1].value))
+        symbol = symbol + x + 'f';
```
Comment (LaykDimon) [inline packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts:68]: maybe put 'f' in const too as use it multiple times?

File: packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+  endModification: DG.InputBase,
+  firstPtoExist: boolean): string {
+  let i: number = -1;
+  let mainSequence = sequence.replace(/[AUGC]/g, function(x: string) {
+    i++;
+    const AXOLABS_MAP = axolabsStyleMap;
+
+    const baseChoices: string[] = Object.keys(AXOLABS_MAP);
+    // const defaultBase: string = baseChoices[0];
+    const indexOfSymbol = AXOLABS_MAP['RNA']['symbols'].indexOf(x);
```
Comment (LaykDimon) [inline packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts:64]: we can put these strings to const

File: packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+    let symbol = AXOLABS_MAP[bases[i].value]['symbols'][indexOfSymbol];
+    if (isOverhang(bases[i].value)) {
+      if (i < sequence.length / 2 && !isOverhang(bases[i + 1].value))
+        symbol = symbol + x + 'f';
+      else if (i > sequence.length / 2 && !isOverhang(bases[i - 1].value))
+        symbol = x + 'f' + symbol;
+    }
+    return (ptoLinkages[i].value) ? symbol + 's' : symbol;
+  });
+  if (mainSequence.slice(0, 5).split('mU').length == 3)
```
Comment (LaykDimon) [inline packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts:74]: strict comparison

File: packages/SequenceTranslator/src/model/format-translation/format-converter.ts [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    const regex = this.buildRegex(Object.keys(codeMapping));
+    return this.sequence.replace(regex, (code) => codeMapping[code]);
+  }
+
+  private bioSpringToGcrs(codeMapping: KeyToValue): string {
+    let count: number = -1;
+    return this.sequence.replace(this.buildRegex(Object.keys(codeMapping)),
+      function(x: string) {
+        count++;
+        return (count == 4) ? codeMapping[x].slice(0, -3) + 'ps' : (count == 14) ? codeMapping[x].slice(0, -2) + 'nps' : codeMapping[x];
```
Comment (LaykDimon) [inline packages/SequenceTranslator/src/model/format-translation/format-converter.ts:51]: put numbers into const

Comment (LaykDimon) [reply packages/SequenceTranslator/src/model/format-translation/format-converter.ts:51]: also strict comparison

**Comment by LaykDimon** [comment]
many of the notes here apply to the whole code, so I recommend you to check it again

File: packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts (hunk 16 above)
Comment (LaykDimon) [reply packages/SequenceTranslator/src/model/axolabs/axolabs-tab.ts:68]: you're welcome!

File: packages/Bio/src/utils/monomer-lib.ts [hunk 20]
```diff
@@ -41,55 +47,55 @@ export async function setUserLibSetting(value: LibSettings): Promise<void> {
 }
 
 export class MonomerLib implements IMonomerLib {
-  private _monomers: { [type: string]: { [name: string]: Monomer } } = {};
+  private _monomers: { [polymerType: string]: { [monomerSymbol: string]: Monomer } } = {};
```
Comment (StLeonidas) [inline packages/Bio/src/utils/monomer-lib.ts:50]: create additional structure

File: packages/Bio/src/utils/monomer-lib.ts [hunk 21]
```diff
@@ -41,55 +47,55 @@ export async function setUserLibSetting(value: LibSettings): Promise<void> {
...
-  constructor(monomers: { [type: string]: { [name: string]: Monomer } }) {
+  constructor(monomers: { [polymerType: string]: { [monomerSymbol: string]: Monomer } }) {
     this._monomers = monomers;
   }
 
-  getMonomer(monomerType: string, monomerName: string): Monomer | null {
-    if (monomerType in this._monomers! && monomerName in this._monomers![monomerType])
-      return this._monomers![monomerType][monomerName];
+  getMonomer(polymerType: string, monomerSymbol: string): Monomer | null {
+    if (polymerType in this._monomers! && monomerSymbol in this._monomers![polymerType])
```
Comment (StLeonidas) [inline packages/Bio/src/utils/monomer-lib.ts:58]: Throw exception and handle it outside. Return Monomer , not Monomer | null

File: packages/SequenceTranslator/src/model/registration/sequence-parser.ts [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    let result = this.prepareInput(strands).slice(DIMER_PREFIX.SS.length).split(DIMER_PREFIX.AS1);
+    const as = this.prepareStrand(result[0]);
+    result = result[1].split(DIMER_PREFIX.AS2).map((strand) => this.prepareStrand(strand));
+    const as1 = result[0];
+    const as2 = result[1];
+    return {ss: as, as1: as1, as2: as2};
+  }
+
+  // todo: port validation to dedicated class
+  private prepareInput(strands: string) {
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/registration/sequence-parser.ts:29]: Always annotate return types

File: packages/SequenceTranslator/src/model/parsing-validation/sequence-validator.ts [hunk 23]
```diff
@@ -0,0 +1,52 @@
+import {NUCLEOTIDES} from '../const';
+import {MonomerLibWrapper} from '../monomer-lib/lib-wrapper';
+import {sortByReverseLength} from '../helpers';
+import {SYNTHESIZERS as FORMAT} from '../const';
+
+export class SequenceValidator {
+  constructor(private sequence: string) {
```
Comment (skalkin) [inline packages/SequenceTranslator/src/model/parsing-validation/sequence-validator.ts:7]: I'm not sure this is the best pattern. Why is it a class and not just a couple of methods? Or something like this, if you really need it:

const validator = new SequenceValidator(format);
validator.validate(sequence
);
